### ISSUE TITLE
[bug] Fibaro Button not working on 2.0.6 (regression from 2.0.5)

### ISSUE THREAD

**Issue by RobertMe** [OpenZWave/Zwave2Mqtt#170]
**Version**

Build/Run method
- Docker

Zwave2Mqtt version: 2.0.6
Openzwave Version: ? (What's bundled with the docker image)

**Describe the bug**
When starting to use ZWave2Mqtt a couple of weeks ago I could add my Fibaro Button just fine. It showed up in the nodes list with type, product and all and I was able to define a group / association to control a Fibaro Wall Plug just fine (so pressing the button would directly send an on/off command to the Wall Plug without intervention of the controller).
Now after some fiddling around, resetting the (hardware) controller and starting with a clean "store" for the Docker container I couldn't get the Button added anymore. It would be added as a node, but it was shown as dead, didn't have a type nor product, couldn't define any groups/associations for it (node is listed on the Groups tab, but then doesn't have any possible groups).
After lots of trial and error of resetting the button, resetting the controller, removing the storage of the docker container I still couldn't get it to work. With ozw debug logging enabled I couldn't see any odd messages.
In the end I checked if there still were some old Docker images and found there apparently aren't any on Docker Hub (which IMO is a shame, not sure if you're manually removing them?). But luckily I still had one locally. Started this image instead, re-added the Button and it immediately showed up just fine with type, product and all. So apparently there is some issue with the latest version or Docker image.

**To Reproduce**
1. Make sure to use the latest docker image (which according to a docker inspect is `robertslando/zwave2mqtt@sha256:78ebac7358cf0bac4fc76ed43906d65f997bd39e64edd8d0d592245fc25954ce` which I have on my RPi)
2. Use "Add node"
3. Activate inclusion on the Fibaro Button
4. Wait for a bit after the "Controller status: Completed" and then refresh the page

**Expected behavior**
The Fibaro Button is property listed in the nodes, with type: `Wall controller` and product `FGPB101 Button (FIBARO System)`. Also when going to Groups and selecting the node the "Group" dropdown is filled (lifeline, on/off, dimmer, alarm). But none of this is the case. (Type and product are empty and Group shows an empty list)

**Additional context**
Using an older docker version, being `robertslando/zwave2mqtt@sha256:b429a9e1dfbbf3a410238281b9013dcf05130a7d3725ded2cc5a1e479c301a89`, this does work fine.

**Comment by robertsLando** [comment]
Were you using the latest tag? 

docker pull robertslando/zwave2mqtt:latest ?

If so Could you try with the dev tag?

docker pull robertslando/zwave2mqtt:2.0.6-dev


Daniel

**Comment by robertsLando** [comment]
Anyway until version 2.0.5 only latest image where keeped on docker hub, this because versioning in dockerhub is a little bit confusing when providing images for different archs. 

Anyway starting from version 2.0.6 now all version will be keeped and there is also a new dev tag that will use ozw 1.6, the non-dev tags will have ozw 1.4 inside. 

Daniel

**Comment by RobertMe** [comment]
Thanks for the quick reply :)

Yes, the broken image corresponds to `latest`. I wrote them down by digest as those are unique and permanent. So when latest gets updated the old digest still uses the old image.

Just tried `2.0.6-dev` and that works fine (think it requires a remove => add cycle though). So this would be a regression in the ozw 1.4 library or how it's installed? As I seem to remember you updated the installation method recently.

**Comment by robertsLando** [comment]
Yes I think it might be something related to ozw that’s why I suggested you to use dev version, that has most recent support for all devices and should work in most cases, it is still “dev” because not all ozw 1.6 features are fully supported by underlying library ozw-shared (nodejs wrapper of ozw). 

Daniel

**Comment by RobertMe** [comment]
But IIRC the old docker image, which also does work, also uses OZW 1.4? So then it would be an regression in one of the more recent OZW 1.4 versions which should be reported there as well.

**Comment by robertsLando** [comment]
Yes it uses 1.4 too and yes could be something broken there 

Daniel

**Comment by petergebruers** [comment]
OZW contributor here... A "Fibaro Button" uses "Command Class Central Scene". Central Scene capabilities are automatically discovered in 1.6, with any device, out of the box. Central Scene in 1.4 is broken and the workarounds you can find online are a hack. @Fishwaldo has stated, there are no plans to port back CentralScene of 1.6 to 1.4, that would defeat the point of having 1.6...

**Comment by RobertMe** [comment]
@petergebruers I've read that there is a lot not working with scenes (and thus the Button) on OZW 1.4. But for my use case it would be fine if the button just showed up and I could define the group association (so it can directly control the Wall Plug). So for this the Button only has to be recognized properly, which does work on the Docker build with (Z2M) 2.0.5, but doesn't anymore in the Docker build with 2.0.6. So I believe that there is either a regression in how OZW is installed in the Docker image (as I believe I've read something about device configuration files being used incorrectly (OZW 1.6 device config files being used on top op OZW 1.4 IIRC)), or there is a regression in OZW 1.4.

So I'm fine by the "scenes will never work on OZW 1.4" but it would be nice if I at least could configure the Button again without having to use the "dev" builds of Z2M.

So to summarize:
Z2M 2.0.5 docker image with OZW 1.4: Button shows in nodes list as sleeping, type and product are listed as well, and group associations can be defined

Z2M 2.0.6 docker image with OZW 1.4: Button can be added, but is shown as dead, type and product are missing, and no group associations can be defined

Z2M 2.0.6-dev docker image with OZW 1.6: Button can be added, is listed as Sleeping, type and product are shown and group associations can be defined. (And possibly much more, which I didn't try yet / don't need at the moment)

**Comment by RobertMe** [comment]
By the way, if the conclusion is that it worked "by accident" in the 2.0.5 builds (for example due to device configuration files mixup) then it's fine for me as well.

Just reporting the fact that it did work in the older build. If the conclusion is that it worked due to a bug/"by accident" then it's fine for me as well. But it might just as well be a bug in how the docker image is build or a regression between the OZW 1.4 versions being used in both builds, so that's why I reported it :slightly_smiling_face:

**Comment by petergebruers** [comment]
Thank you for clarifying that. I cannot explain the different behavior between the two 1.4 images from OZW's perspective. It sounds like inclusion wasn''t 100 % successful, the device needed wake up to finish the interview procedure, or its config file was missing, or a combination.

**Comment by RobertMe** [comment]
As I've (tried to) include it lots and lots of times on the 2.0.6 (with OZW 1.4) image and it didn't work a single time, but it does work the first time when trying on the 2.0.5 image and 2.0.6 with OZW 1.6 image I'm pretty sure it's not related to the device not completing the pairing / sleeping to early (pretty familiar with it from the Zigbee side where lots of Xiaomi devices need to be kept awake during pairing by pressing the button). So then I guess it's due to a missing configuration file. @robertsLando might you be able to confirm this? As I'm unfamiliar with how you include these in the image and I seem to remember there were some issues earlier on with 1.6 config files being used on OZW 1.4. (to reiterate: if it previously worked due to the 1.6 files being used in combination with 1.4 then I'm fine by it. As using the wrong config obviously can lead to bugs as well and the correct ones obviously need to be used)

**Comment by Fishwaldo** [comment]
OZW 1.4 hasnt had any changes in ~6 months apart from recently a corner case that dealing with failed startups and modemanager.

I don't know what the difference is between 2.0.5 and 2.0.6 but if it was the config file mixup (using 1.6 config files on 1.4) then that might indicate the confusion

As determined - using 1.6 config files on 1.4, although they load, they "dont" really work as intended and some devices would go haywire.

**Comment by petergebruers** [comment]
I checked, the file for the Fibaro button on 1.4 hasn't changed in 3 years... This is the one for 1.4...

https://github.com/OpenZWave/open-zwave/blob/1.4/config/fibaro/fgpb101.xml

As Fishwaldo points out... A 1.6 config is not backwards compatible, you can recognize the new format by looking at the first line, if it looks like `<Product Revision="6" xmlns="https://github.com/OpenZWave/open-zwave">` then it is in 1.6 format

**Comment by RobertMe** [comment]
Hmm. I've just extracted the fgpb.xml file from both Docker images (so 2.0.5 and 2.0.6 both with OZW 1.4) and the files are identical. So that can't be the issue.

So next question would then be: what else can be the issue? Possible the OZW wrapper for NodeJS, or obviously ZWave2Mqtt itself but I don't think that can have any real influence on the pairing procedure?

**Comment by petergebruers** [comment]
I would try to compare detailed logging of OZW. Enable "OpenZWave debug logging" if that is not the default on Zwave2Mqtt. Also wake up the button (click the Button 4 times). Then have a look at OZW_Log.txt. If you cannot spot any difference, I'll have a look those files (but be patient, I am slow...). Just zip and drag and drop them here.

**Comment by RobertMe** [comment]
I'm currently quickly reading the log for the working version and one thing I immediately notice is the line:
```
Opening config param file /usr/local/etc/openzwave/fibaro
/fgpb101.xml
```
As I couldn't remember reading something like this yesterday in the log for the non working version. And indeed this log line is missing.

So could you maybe give me some quick pointers what triggers the reading of the config so I can do some better comparison of the logs? I presume OZW (or the stick) sends some message to get the device information and based on the response reads the config file. So based on for example that message I can continue my investigation to which parts are and aren't present in the broken version.

**Comment by RobertMe** [comment]
So checking more of the log I believe this is where it goes wrong:
```
2019-12-11 14:24:26.990 Detail, Node015, QueryStage_WakeUp
2019-12-11 14:24:26.991 Detail, Node015, Queuing (WakeUp) WakeUpCmd_IntervalGet (Node=15): 0x01, 0x09, 0x00, 0x13, 0x0f, 0x02, 0x84, 0x05, 0x25, 0x33, 0x7f
2019-12-11 14:24:26.992 Detail, Node015, Queuing (Query) Query Stage Complete (WakeUp)
2019-12-11 14:24:26.992 Detail,
2019-12-11 14:24:26.992 Info, Node015, Sending (WakeUp) message (Callback ID=0x33, Expected Reply=0x04) - WakeUpCmd_IntervalGet (Node=15): 0x01, 0x09, 0x00, 0x13, 0x0f, 0x02, 0x84, 0x05, 0x25, 0x33, 0x7f
2019-12-11 14:24:27.001 Detail, Node015,   Received: 0x01, 0x04, 0x01, 0x13, 0x01, 0xe8
2019-12-11 14:24:27.001 Detail, Node015,   ZW_SEND_DATA delivered to Z-Wave stack
2019-12-11 14:24:27.021 Detail, Node015,   Received: 0x01, 0x18, 0x00, 0x13, 0x33, 0x00, 0x00, 0x02, 0x00, 0xbc, 0x7f, 0x7f, 0x7f, 0x7f, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x03, 0x01, 0x00, 0x00, 0x78
2019-12-11 14:24:27.022 Detail, Node015,   ZW_SEND_DATA Request with callback ID 0x33 received (expected 0x33)
2019-12-11 14:24:27.022 Info, Node015, Request RTT 30 Average Request RTT 47
2019-12-11 14:24:27.023 Detail,   Expected callbackId was received
2019-12-11 14:24:36.993 Error, Node015, ERROR: Dropping command, expected response not received after 1 attempt(s)
2019-12-11 14:24:36.993 Detail, Node015, Removing current message
2019-12-11 14:24:36.994 Detail, Node015, Notification: Notification - TimeOut
2019-12-11 14:24:36.994 Detail, Node015, Query Stage Complete (WakeUp)
```

So after 10 seconds a timeout is reported. While the working version continues on:

```
2019-12-11 13:16:06.903 Detail, Node014, QueryStage_WakeUp
2019-12-11 13:16:06.904 Detail, Node014, Queuing (WakeUp) WakeUpCmd_IntervalGet (Node=14): 0x01, 0x09, 0x00, 0x13, 0x0e, 0x02, 0x84, 0x05, 0x25, 0x31, 0x7c
2019-12-11 13:16:06.904 Detail, Node014, Queuing (Query) Query Stage Complete (WakeUp)
2019-12-11 13:16:06.904 Detail,
2019-12-11 13:16:06.905 Info, Node014, Sending (WakeUp) message (Callback ID=0x31, Expected Reply=0x04) - WakeUpCmd_IntervalGet (Node=14): 0x01, 0x09, 0x00, 0x13, 0x0e, 0x02, 0x84, 0x05, 0x25, 0x31, 0x7c
2019-12-11 13:16:06.913 Detail, Node014,   Received: 0x01, 0x04, 0x01, 0x13, 0x01, 0xe8
2019-12-11 13:16:06.914 Detail, Node014,   ZW_SEND_DATA delivered to Z-Wave stack
2019-12-11 13:16:06.933 Detail, Node014,   Received: 0x01, 0x18, 0x00, 0x13, 0x31, 0x00, 0x00, 0x02, 0x00, 0xc6, 0x7f, 0x7f, 0x7f, 0x7f, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x03, 0x01, 0x00, 0x00, 0x00
2019-12-11 13:16:06.934 Detail, Node014,   ZW_SEND_DATA Request with callback ID 0x31 received (expected 0x31)
2019-12-11 13:16:06.934 Info, Node014, Request RTT 28 Average Request RTT 28
2019-12-11 13:16:06.934 Detail,   Expected callbackId was received
2019-12-11 13:16:06.948 Detail, Node014,   Received: 0x01, 0x0e, 0x00, 0x04, 0x00, 0x0e, 0x06, 0x84, 0x06, 0x00, 0x00, 0x00, 0xff, 0xc6, 0x00, 0x46
2019-12-11 13:16:06.948 Detail,
2019-12-11 13:16:06.948 Info, Node014, Response RTT 42 Average Response RTT 42
2019-12-11 13:16:06.949 Info, Node014, Received Wakeup Interval report from node 14: Interval=0, Target Node=255
2019-12-11 13:16:06.950 Detail, Node014, Initial read of value
2019-12-11 13:16:06.950 Detail, Node014, Queuing (WakeUp) WakeUpCmd_IntervalSet (Node=14): 0x01, 0x0d, 0x00, 0x13, 0x0e, 0x06, 0x84, 0x04, 0x00, 0x00, 0x00, 0x01, 0x25, 0x32, 0x7f
2019-12-11 13:16:06.950 Detail, Node014,   Expected reply and command class was received
2019-12-11 13:16:06.951 Detail, Node014,   Message transaction complete
2019-12-11 13:16:06.951 Detail,
2019-12-11 13:16:06.951 Detail, Node014, Removing current message
2019-12-11 13:16:06.952 Detail, Node014, Notification: ValueChanged
2019-12-11 13:16:06.952 Detail,
2019-12-11 13:16:06.953 Info, Node014, Sending (WakeUp) message (Callback ID=0x32, Expected Reply=0x13) - WakeUpCmd_IntervalSet (Node=14): 0x01, 0x0d, 0x00, 0x13, 0x0e, 0x06, 0x84, 0x04, 0x00, 0x00, 0x00, 0x01, 0x25, 0x32, 0x7f
2019-12-11 13:16:06.962 Detail, Node014,   Received: 0x01, 0x04, 0x01, 0x13, 0x01, 0xe8
2019-12-11 13:16:06.962 Detail, Node014,   ZW_SEND_DATA delivered to Z-Wave stack
2019-12-11 13:16:06.982 Detail, Node014,   Received: 0x01, 0x18, 0x00, 0x13, 0x32, 0x00, 0x00, 0x02, 0x00, 0xc6, 0x7f, 0x7f, 0x7f, 0x7f, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x03, 0x01, 0x00, 0x00, 0x03
2019-12-11 13:16:06.982 Detail, Node014,   ZW_SEND_DATA Request with callback ID 0x32 received (expected 0x32)
2019-12-11 13:16:06.982 Info, Node014, Request RTT 29 Average Request RTT 28
2019-12-11 13:16:06.982 Detail,   Expected callbackId was received
2019-12-11 13:16:06.983 Detail,   Expected reply was received
2019-12-11 13:16:06.983 Detail,   Message transaction complete
2019-12-11 13:16:06.983 Detail,
2019-12-11 13:16:06.983 Detail, Node014, Removing current message
2019-12-11 13:16:06.983 Detail, Node014, Query Stage Complete (WakeUp)
```

More precisely, the (first) response differs
```
Broken:
0x01, 0x18, 0x00, 0x13, 0x33, 0x00, 0x00, 0x02, 0x00, 0xbc, 0x7f, 0x7f, 0x7f, 0x7f, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x03, 0x01, 0x00, 0x00, 0x78
Working:
0x01, 0x18, 0x00, 0x13, 0x31, 0x00, 0x00, 0x02, 0x00, 0xc6, 0x7f, 0x7f, 0x7f, 0x7f, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x03, 0x01, 0x00, 0x00, 0x00
```
So except the obvious difference (different callback id of 0x33 vs 0x31) the tenth byte differs (0xbc vs 0xc6) and the last byte differs (0x78 vs 0x00). But obviously I don't know the meaning of the message, let alone these specific bytes.

**Comment by petergebruers** [comment]
I am not at home, will have to look at in more detail later, but I'd say the log with the timout indeed means the "interview" process does not compleet, the node should reply. Because this is the same hardware and same OZW version that is odd. You say this is reproducible, but this thought keeps nagging at me: some versions, probably the early proluction ones, of the "button" are known to have a hardware issue: the springy part (in the "head of the mushroom") can wear out, leading to power supply issues. This causes weird behaviour. Most people describe it like this: it is as if the devices goes to some deep sleep, and requires several button presses before it seems to work again. Mine gave issues after several months. I tried do bend that spring, but it broke. Then I "repurposed" the PCB and it works just fine. It purely failing mechanics...

**Comment by RobertMe** [comment]
I've had two of them as well. Had to return the first one as it didn't work (so probably the same hardware issue).

But this one seems to work fine, although I haven't used it in ages (want to use it know for the Christmas tree now that I have a spare plug as well :stuck_out_tongue: ).

But you say the interview doesn't work as the button doesn't reply. But from how I read the log the button does reply. But most likely with a different (unhandled?) response.

**Comment by Fishwaldo** [comment]
The last bytes in each message are CRC so they will change. 

Does it *always* fail on the WakeUp_interval message in 2.0.5?

**Comment by RobertMe** [comment]
> The last bytes in each message are CRC so they will change.

That indeed does explain the difference.

> Does it always fail on the WakeUp_interval message in 2.0.5?

It's the other way around. 2.0.5 _seemed_ to work while 2.0.6 didn't.

And yes _seemed_. Yesterday I tried again with both versions to gather some logs. But I even couldn't get it paired property with 2.0.5 anymore. For both the device was added as dead. At some point the Button wouldn't even get paired at all (controller stayed in 'Waiting' status while normally it would immediately return to 'Completed' after enabling the pairing mode on the Button). In the end I did manage to get it paired again but I'm not sure how (I at least did a reset of the Button after which it did respond to the pairing again, but not sure what I did to get the pairing fully completed). So I'll think I'll just leave it for now and maybe the Button indeed is broken one way or another. (As said, didn't use it in maybe even more than three years. So not sure if it did work properly and every single time back then, and that even was in combination with an Athom Homey controller so not comparable with my currect ZWave.me stick with OZW)

I do have two extra logs of 2.0.6 though (from before I found out I couldn't get it paired on 2.0.5 either) and it does seem that in both those cases it also failed during the WakeUp stage.
Sadly I didn't copy the logs after I eventually did get it paired and I did a restart later on so I don't have that log anymore. So I don't know how it did complete that time (maybe just lucky with timing).

**Comment by petergebruers** [comment]
IMHO it is a hardware issue, it is exactly what happened to my "Button"... I'll try to find the posts on the Fibaro Forum, someone posted a video about fixing that contact. You still have warranty?

**Comment by petergebruers** [comment]
This post is from a reseller, he says a certain % is failing and when that happens weird things happen eg with battery status report...

https://forum.fibaro.com/topic/32737-button-unreliable-strange-zwave-protocol-responses

Then there is this long topic, this link leads you to the :

https://forum.fibaro.com/topic/26679-so-is-the-fibaro-button-working-for-anybody

In that topic there is a link to this video, showing how a user fixed that contact.

https://youtu.be/Xt4HYeD0AAY

Warning, In may case, that part broke off, I repurposed the PCB and it working fine, so it is not a firmware or a device issue.

**Comment by RobertMe** [comment]
Thanks for the references @petergebruers . But I'm now getting more sure that it isn't a hardware/battery issue. For one pairing and unpairing are reliable again, 6 clicks, and the controller recognizes it, although it might not _properly_ pair (so shown as dead without product info). But most importantly, when using a group association it is (in a small test) 100% reliable, and every press changes the state of the associated wall plug.

But, most importantly, after some trial and error I know have it paired on the ZWave2Mqtt 2.0.6 docker image. So there at least is no issue with Z2M.
And I believe that pairing is relatively reliable when doing the 6x click to start the pairing, waiting a bit, clicking 4 times for a wake up, waiting about half a second to a second again, once more click 4 times, waiting again to do a wake up, ....

So closing as it's at least not an issue with Z2M / the docker image. (If desired I might continue "debugging" to find out if the issue is in OZW. But I don't presume that's the case and as OZW 1.4 isn't activally maintained I guess you're not interested in it either)

**Comment by petergebruers** [comment]
Thank you for your continued effort. It is possible your "click rythm" is special, though generally speaking doing the "wake up sequence" one or more times should do the trick. You are correct when you say OZW contributors won't investigate this for 1.4, unless it would be some really fundamental and annoying issue.

**Comment by petergebruers** [comment]
I think it worth pointing out that OZW 1.6, in general, "auto clears" these values:
Scene Activation
Central Scene
Alarm aka Notification

When one of those events are reported by the device, OZW will change the value to indicate it got triggered. Then, after some time OZW will clear that value.

So In the log you first see the button press or alarm, then after some time:

""Automatically Clearing Scene" if it is Central Scene
""Automatically Clearing SceneID/Duration in" if it is Scene Activation
""Automatically Clearing Alarm in %dms"" if is an Alarm (Notificaton)

In case of the Alarm, the situation is more complex than the Scene classes and here is a separate discussion about that:

https://github.com/OpenZWave/open-zwave/issues/1889 "COMMAND_CLASS_ALARM "magic" clear issue"
